README.md
# Ev3
C++ symbolic manipulation libraries for mathematical expressions

This C++ library was written for the Reformulation-Optimization Software Engine (ROSE) <https://github.com/coin-or/ROSE/>, which is a software for performing symbolic reformulations on Mathematical Programming formulations. While ROSE was the main output of an ANR project (JCJC2007, "ARS"), Ev3 was written in 2005-2006. Both Ev3 and ROSE are scarcely maintained since 2012, so you are welcome to use them at your own peril. They worked reasonably well with compilers and hardware of their time. 


HISTORY
# history file for Expression v3 project
# RCS1 "$Id: HISTORY,v 1.21 2006/07/30 05:36:27 liberti Exp liberti $"

010517	work started

010518	basic classes seem to be working.

010518	no makefile for now, compile with c++ -o main main.cxx

010622	working on PlusLink - just started, copied it from PlusCopy,
	have to modify it yet.

010624	adapted the parser from the Eval project to work with
	Expression v3.

010721	porting to gcc 2.7.2.3, sigsegvs at parser.cxx:228

010722	realized an IMMENSE mistake - we have to use Pointer<Expression>,
	not just Expression, to enable to automatic garbage collector.
	Have to revolutionize everything. Rename Expression to
	BasicExpression and typedef Pointer<BasicExpression> Expression;
	Before attempting such a big change, I am going to check everything
	in.

010722	fixed the IMMENSE mistake - architecture has changed, though.
	Have to update the document (article).
	1. public methods for constructions are OUTSIDE of classes
	2. they are: operator+,-,*,/,^,Log,Exp,Sin,Cos,Tan,Sinh,...,Sqrt
	3. they all work by changing the arguments, so in case you
	   don't want it changed, use things like 
		Expression e = f.Copy() + g.Copy();
	   instead of
		Expression e = f + g;
	4. Parser uses linking (referencing), not copying. Realized
	   it makes more sense because the "bricks" left around would
	   be deleted anyway.

010722	Diff seems to be working. Needs more testing though. Can
	probably finish off the article now.

010723	Put some simplification in place.

010723	example TrigSimp is working

021122	Added stuff for use with MORON

021209	CHECK IN --- first in the MORON usage

021218	Should write a little command-line routine that simplifies
	and diffs equations - so that I can test everything.

021218	There should be a routine implementing "distribute sum over products":
	it is convenient for subsequent convexifications (besides,
	cases like 2(x + y + z) should really be transformed into
	2x + 2y + 2z for linearity's sake).

021218	Now the parser automatically allocates varindices if it
	is not initialized manually.

021218	CHECK IN

021219	Problems in:
	- simplification of fractions;
	- y+x-0.5*y-x-0.5*y SIGSEGVs on Expression zero(0.0): fishy

021220	FISHY PROBLEM IN 021219 SOLVED EASILY!!! It's the first time
	I manage to dispose of a nasty, random, tree-related SIGSEGV
	so effectively. Valgrind helped me ascribe the bug to a
	--(*ncount) in tree.cxx where ncount was not allocated: and
	since ncount only appears in the short tree.cxx file, it was
	then easy to insert a ncount = NULL all the time where
	I'd forgotten it. Et voila`.

021220	simplification of fractions now works OK

021220	noticed Diff misses cases COT, SINH, COSH, TANH, COTH, SQRT

021220	CHECK IN

030106	GetLinearInfo() and GetNonlinearPart() separate the linear
	and nonlinear parts of an expression. However, watch out!
	In an expression like x+y+y^2 the linear part is defined
	as x and the nonlinear as y+y^2. This is against the mathematical
	common sense (which would require x+y as linear and y^2 as
	nonlinear part) but it makes it easier to Moron to interface
	to most NLP local solvers.

030114	Working on Interval()

030115	Have to do powermkrange() properly, and fill in the missing
	gaps in Interval() (hyperbolic functions)

030116	changed operators so that they never modify their arguments.
	Now, the parser relies on the old model, with modification
	of arguments. Therefore: take all the old operators,
	rename them suitably, and change the parser to use the old
	model operators.

030116	The behaviours of operator+, operator-, operator*, ...
	and SumLink, DifferenceLink, ProductLink, ... is similar
	yet very different. The fact that the "Link" version of
	the call changes the arguments and minimizes the number of
	created subnodes makes it very useful to create expressions
	from scratch (as in the parser), where the constituent nodes
	do not have an independent life. When creating a new
	constraint from an existing one - and the existing one has, 
	as it were, a life of its own, then the "operator" version
	of the calls should be used. I have also observed that
	the stack does not handle temporary Expression entities
	graciously; thus, a C++ instruction like
		do_something(a + b / c);
	where a,b,c are Expressions, might fail or give unexpected
	results. Use with care, always test first. If in doubt do
	one step at a time:
		Expression t1 = a + b;
		Expression t2 = t1 / c;
		do_something(t2);
	Things like 
		Expression e = something;
		Expression f = something_else;
		e = e + f;
	might not work. Use SumLink() instead, as in e = SumLink(e, f).

030117	Added "AddCopyOfNode()" to tree.cxx, for use in the 
	operator+,*,... (i.e. in the non -Link version of the
	construction methods).

030117	Changing expression.h:LARGE from 1e10 to 1e30 has the
	effect of unearthing a strange SIGSEGV in DiffNoSimplify()
	due to an exceptions there being no nodes, called from
	expression.cxx:3386. Why?

030117	CHECK IN

030121	Bugfix to go with the convexification routines

030121	Big bugfix to IsSmithStandard (now depends on
	constraint bounds as well)

030121	CHECK IN

030122	Ported to GCC 3.2 (stringstream instead of strstream
	and things like that. Ah: stringstream doesn't like
	<< ends).

030914	Can now compile with both GCC < 3 and >= 3

030925	Now simplifies (var)^(const) in var with a constant
	exponent (SimplifyRecursive)

030925	Remember that there are WARNINGs scattered around, mostly

expression.h
//                                               -*- C++ -*-
/**
 *  @brief symbolic expression (base classes and functionality)
 *
 *  Copyright (C) 2008-2010 Leo Liberti
 *
 *  This library is free software: you can redistribute it and/or modify
 *  it under the terms of the GNU Lesser General Public License as published by
 *  the Free Software Foundation, either version 3 of the License, or
 *  (at your option) any later version.
 *
 *  This library is distributed in the hope that it will be useful,
 *  but WITHOUT ANY WARRANTY; without even the implied warranty of
 *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *  GNU Lesser General Public License for more details.
 *
 *  You should have received a copy of the GNU Lesser General Public License
 *  along with this library.  If not, see <http://www.gnu.org/licenses/>.
 *
 */

#ifndef __EV3EXPRESSIONH__
#define __EV3EXPRESSIONH__

#define RCS5 "$Id: expression.h,v 1.25 2006/07/30 05:36:41 liberti Exp liberti $"

using namespace std;
#ifdef SUNWIN
#include <iostream>
#endif

#include <vector>
#include <string>
#include <map>

////////////// MEMCHECK debug ////////////////////
#ifdef MEMDEBUG
extern vector<pair<void*,pair<int,int*> > > memcheckdebug;
extern vector<pair<void*,pair<int,int*> > >::iterator memcheckdebugit;
extern pair<void*, pair<int, int*> > memcheckdebugpair;
extern int memcheckdebugcounter;
#endif
///////////// END MEMCHECK debug ////////////////

#include "common.h"
#include "tree.cxx"
#include "exceptions.h"
#include "fastexpression.h"

// algebraic expression operand
class Operand {

private:

protected:

  // one of the OperatorTypes above
  int oplabel;

  // 0 if no dependency, 1 if constant, 2 if coefficient, 3 if exponent
  int dependency;

  // if oplabel == CONST, the value of the constant
  double constant;

  // dependency for constants (added for MORON - see ../PROGNOTES)
  double* depconstant;

  // if oplabel == VAR, the index of the variable - should start from 1
  Int varindex;

  // if oplabel == VAR, the name of the variable
  string varname;

  // we allow multiplication for a constant coefficient in each Operand
  double coefficient;
  
  // dependency for coefficients (added for MORON - see ../PROGNOTES)
  double* depcoefficient;

  // we allow a real constant exponent in each Operand
  // THIS HAS MEANING ONLY IF operand IS A LEAF!!!
  double exponent;

  // dependency for exponents (added for MORON - see ../PROGNOTES)
  double* depexponent;

  // interval associated to the values of this node (used in FBBT)
  double lb;
  double ub;

public:

  // constructors
  Operand();
  Operand(const double t);
  Operand(const Int t);
  Operand(const Int t, const bool isvar);
  // create a variable leaf and set coefficient
  Operand(const double c, const Int t, string vn);

  // copy constructor
  Operand(const Operand& t);

  // destructor
  ~Operand();

  // Operand class methods:

  // prints to a string
  string ToString(void) const;

  // get operator type
  int GetOpType(void) const;
  
  // get constant value - in CONSTs it multiplies by coefficient and
  // raises to exponent
  double GetValue(void) const;

  // just get the value, in all cases
  double GetSimpleValue(void) const;

  // get variable index
  Int GetVarIndex(void) const;

  // get variable name
  string GetVarName(void) const;

  // get the coefficient
  double GetCoeff(void) const;

  // get the exponent
  double GetExponent(void) const;
  
  // get the lower bound
  double GetLB(void) const;
  
  // get the upper bound
  double GetUB(void) const;

  // set operator type
  void SetOpType(const int t);
  
  // set constant value
  void SetValue(const double t);  

  // set variable index (start from 1 and add by steps of 1 when creating new
  // variables)
  void SetVarIndex(const Int t);

  // set variable name
  void SetVarName(const string vn);

  // set the exponent
  void SetExponent(const double expon);

  // set the coefficient
  void SetCoeff(const double coeff);

  // set the lower bound
  void SetLB(double);
  
  // set the upper bound
  void SetUB(double);

  // set constant dependencies (added for MORON - see ../PROGNOTES)
  void SetDependencyOnOperand(const int whichconstant, double** depvalue);

  // is operand a constant?
  bool IsConstant(void) const;

  // is operand a variable?
  bool IsVariable(void) const;

  // is operand a leaf node?
  bool IsLeaf(void) const;

  // is operand a zero constant?
  bool IsZero(void) const;
  
  // is operand a constant == v?
  bool HasValue(double v) const;

  // is operand a constant <= v?
  bool IsLessThan(double v) const;

  // is operand constant >= v?
  bool IsGreaterThan(double v) const;

  // set value = coefficient * value ^ exponent
  void ConsolidateValue(void);

  // enforce constant dependencies (added for MORON - see ../PROGNOTES)
  void EnforceDependencyOnOperand(void);

  // is operand this == operand t?
  bool operator == (const Operand& t);

  // substitute a variable with a constant
  void SubstituteVariableWithConstant(int varindex, double c);  

};

class BasicExpression;

typedef Pointer<BasicExpression> Expression;

class BasicExpression : public Operand, public Tree<BasicExpression> {

private:
  FastEvalTree* fetree;
  bool FastEvalAllocated;

public:

  // constructors

  // create empty
  BasicExpression();

  // create a constant leaf
  BasicExpression(const double t);

  // create a constant (integer-valued) leaf
  BasicExpression(const Int t);

  // create an (empty) operator or a variable leaf
  BasicExpression(const Int t, const bool isvar);
  
  // create a variable leaf and set coefficient
  BasicExpression(const double c, const Int t, string vn);

  // user-defined copy constructor with two options (to make a copy)
  BasicExpression(const Expression& t, const bool iscopy);

  // copy constructor
  BasicExpression(const BasicExpression& t);

  // destructor
  ~BasicExpression();

  // BasicExpression class methods:
  void Debug (void) const;

  // prints to a string
  string ToString(void) const;

  // output is a tree
  string PrintTree(int blanks, int tabs) const;

  // sets an expression to zero (deleting all existing subnodes)
  void Zero(void);

  // sets an expression to one (deleting all existing subnodes)
  void One(void);

  // is expression this == expression t?
  // (note that this half-replicates Tree::operator==,
  //  but I couldn't think of any other convenient way to fit in
  //  the operand data in == and still use the Tree's ==)
  bool IsEqualTo(const Expression& t) const;
  bool operator == (const BasicExpression& t) const;

  // other comparison operators
  bool IsEqualTo(const double t) const;

  // this returns true if args are equal up to top node coefficient
  bool IsEqualToNoCoeff(const Expression& t) const;

  // this returns true if args have equal operator structures and
  // maintain variable/constant assignment to leaf nodes
  // (a schema is an expression modulo leaves)
  bool IsEqualBySchema(const Expression& t) const;

  // this returns true if args have equal operator label
  bool IsEqualByOperator(const int theoplabel) const;

  // this returns the number of variables in the expression
  int NumberOfVariables(void) const;
  int NumberOfVariables(int& maxvi) const;

  // fast evaluation tree handling
  // creation (use nonrecursive version)
  void CreateFastEvalTree(void);
  void CreateFastEvalTreeRecursive(FastEvalTree* fet);
  // deletion (use nonrecursive version)
  void DeleteFastEvalTree(void);
  void DeleteFastEvalTreeRecursive(FastEvalTree* fet);
  
  // EVALUATION ROUTINES
  // Slow evaluators (for cases when just 1 evaluation is needed;
  // for repeated evaluations, use FastEval outside the class)
  // numeric evaluation: varvalues[i] contains the value of the
  // variable with varindex i, for all i < vsize.
  double Eval(double* varvalues, Int vsize) const;
  // in the following version, the varmap is used to map
  // var. indices in the expressions to indices of varvalues
  double Eval(double* varvalues, map<int,int>& varmap, Int vsize) const;
  // Fast evaluators (for repeated evaluations -- use nonrecursive version)
  double FastEval(double* varvalues, Int vsize);
  double FastEval(double* varvalues, map<int,int>& varmap, Int vsize);  
  double FastEval(FastEvalTree* fet, double* varvalues, Int vsize);
  double FastEval(FastEvalTree* fet, double* varvalues, 
		  map<int,int>& varmap, Int vsize);  
  double FastEvalRecursive(FastEvalTree* fet, double* varvalues, 
			   map<int,int>* varmap, Int vsize);

  // return the FastEvalTree pointer (build it if required)
  FastEvalTree* GetFastEvalTree(void);

  // whether expression depends on variable
  bool DependsOnVariable(Int vi) const;

  // whether expression depends on variable linearly
  // (0=depends nonlinearly, 1=depends linearly, 2=doesn't depend at all)
  int DependsLinearlyOnVariable(Int vi) const;

  // in a product, multiply all coeffs. of the operands, set result
  // as coeff of whole product, reset all operand coffs at 1; if
  // resulting global coeff is zero, delete all nodes and set 
  // this to zero constant.
  // don't do anything on other operators
  void ConsolidateProductCoeffs(void);

  // in a sum or product, if coeff of sum operand is not 1, distribute it
  // to the operands and set whole coeff to 1
  void DistributeCoeffOverSum(void);
  void DistributeCoeffOverProduct(void);

  // enforce constant dependencies (added for MORON - see ../PROGNOTES)
  // this only acts on the proper leaf nodes
  void EnforceDependency(void);

  // substitute a variable with a constant
  void VariableToConstant(int varindex, double c);

  // replace variable indexed v1 with variable indexed v2 (with varname vn)
  void ReplaceVariable(int v1, int v2, string vn);
  void ReplaceVariable(int v1, int v2, string vn, double c2);

  // replace with a variable the deepest node conforming to schema and
  // return replaced term or zero expression if no replacement occurs
  Expression ReplaceBySchema(int vi, string vn, Expression schema);
  // works on subnodes not on current node
  Expression ReplaceBySchemaRecursive(int vi, string vn,Expression schema);

  // replace with a variable the deepest node with given operator label
  // return replaced term or zero expression if no replacement occurs
  Expression ReplaceByOperator(int vi, string vn, int oplabel);  
  // works on subnodes not on current node
  Expression ReplaceByOperatorRecursive(int vi, string vn, int oplabel);  

  // replace all occurrences of subexpression needle with replace
  // return number of replacements
  int ReplaceSubexpression(Expression needle, Expression replace);

  // replace this with given expression (SIGSEGV risk, see implementation)
  void ReplaceWithExpression(Expression replace);

  // smith's standard form: takes a starting added variable index and
  // name and a vector of oplabels and one of schemata, performs replacements
  // on the this expression, returns sequence of defining constraints
  // for a defining constraint w_i = x o y, defcons[i] contains expression
  // x o y; caller must then use this to set constraint w - x o y = 0
  void SmithStandardForm(int defvarindex, string defvarname,
			 vector<int>& oplabels, vector<Expression>& schemata,
			 vector<Expression>& defcons);
  void SmithStandardFormRecursive(int defvarindex, string defvarname,
				  vector<int>& oplabels, 
				  vector<Expression>& schemata,
				  vector<Expression>& defcons);

  // prodbincont: replace products x*y where at least one var is binary
  // with w, and then adds exact linearization constraints for w
  // semantic difference of defcons with SmithStandardForm: here, 
  // defcons contains exactly 4 expressions lin_expr(w,x,y) 
  // to be set <= 0 by caller. Returns number of replacements.
  // addvarbounds contains map from added varindices to corresp var bounds
  int ProdBinCont(int defvarindex, string defvarname,
		  map<int,bool>& integrality,
		  map<int,double>& vlb, 
		  map<int,double>& vub,
		  map<int,pair<double,double> >& addvarbounds,
		  vector<Expression>& defcons);
  // returns number of replacements
  int ProdBinContRecursive(int defvarindex, string defvarname, 
			   map<int,bool>& integrality,
			   map<int,double>& vlb, 
			   map<int,double>& vub,
			   map<int,pair<double,double> >& addvarbounds,
			   vector<Expression>& defcons);

  // reset all names of variables having IDs between lid, uid to vn
  void ResetVarNames(string vn, int lid, int uid);
  
  // distribute products over sums - returns true if changed 
  // (re-call until false)
  bool DistributeProductsOverSums(void);

  // find variable indices in an expression
  void GetVarIndices(vector<int>& vidx);

  // return list of varIDs involved in a certain schema
  // e.g. f(x1*x2+x1*x3+x4*log(x5), x*y) = {1,2,3}
  void GetVarIndicesInSchema(vector<int>& vidx, Expression schema);

  // find the variable name corresponding to variable index vi
  string FindVariableName(int vi);

  // is this expression linear?
  bool IsLinear(void) const;

  // is this expression a quadratic product of variables? 
  // If yes, return the product type: PRODUCT, POWER or VAR
  bool IsQuadratic(int& prodtype) const;
  bool IsQuadratic(void) const;

  // return info about the linear part (assumes Simplify() has already been
  // called on this) - return false if expression has no linear part
  // by "linear part" we mean lin(x) in expr(x,y) = lin(x) + nonlin(y)
  // lincoeff[i] is the i-th nonzero coeff, linvi[i] is the corresponding
  // i-th varindex (starts from 1), c is the constant term to be added
  bool GetLinearInfo(vector<double>& lincoeff, 
		     vector<int>& linvi, vector<string>& linvn, double& c);

  // return info about the pure linear part (assumes Simplify() has
  // already been called on this) - much as above call but the "pure
  // linear part" is e.g. x+y in x+y+y^2
  bool GetPureLinearInfo(vector<double>& lincoeff, 
			 vector<int>& linvi, vector<string>& linvn, double& c);

  // get the linear part - x in x+y+y^2
  Expression GetLinearPart(void);

  // get the pure linar part - x+y in x+y+y^2
  Expression GetPureLinearPart(void);

  // get the nonlinear part - nonlin(y) in expr(x,y) = lin(x) + nonlin(y)
  Expression GetNonlinearPart(void);

  // get the purely nonlinear part - eg. y^2 in x+y+y^2
  Expression GetPureNonlinearPart(void);

  // return the additive constant of the expression 
  double GetConstantPart(void);

  // return the additive constant of the expression and remove it from the
  // expression itself
  double RemoveAdditiveConstant(void);

  // perform interval arithmetics on the expression;
  // Vlb[i] is the lower bound of varindex i, Vub is similar,
  // returned elb and eub hold values to expression interval
  void Interval(map<int,double>& Vlb, map<int,double>& Vub, 
		double& elb, double& eub);

  // feasibility-based bounds tightening: performs one up/down iteration
  // Vlb/Vub input lower/upper bounds for variables, and updated for
  // output; elb/eub: constraint bounds (input only)
  void FBBTUpDown(map<int,double>& Vlb, map<int,double>& Vub, 
		  double elb, double eub);
  void FBBTDown(double elb, double eub);

  // this checks if expression is of the form
  // c * w_i  +/-  (unary_or_binary_operation) 
  // l and u are the constraint's lower and upper bounds
  // where the constraint is l < *this < u - it makes 
  // a difference for likening an inequality or equation
  // to Smith standard form
  bool IsSmithStandard(double l, double u);
  bool IsSmithStandard(void);

  // this checks if the set defined by the constraint 
  // l < *this < u is evidently convex 
  bool IsEvidentlyConvex(double l, double u);
  // this checks if expression *this is evidently convex 
  bool IsEvidentlyConvex(void);
  // this checks if expression *this is evidently concave 
  bool IsEvidentlyConcave(void);

  // this checks if expression is of the form
  // c * w_i  +/-  (operation for which we have a good convexifier) 
  bool IsOptStandard(void);

};

// All these functions contain tricks to simplify the operands. This
// means that both the operands may be changed, and indeed that the
// return value is often one of the changed operands.  To make sure
// you are not changing the operands, use the CopyOf function or the
// Copy() method in Pointer class; or use the equivalent operators below
// in order not to touch the argument expressions.

// add a link of b to a, return link to created expression
Expression SumLink(Expression a, Expression b);
// add a link of b to a with coeff -1, return link of a
Expression DifferenceLink(Expression a, Expression b);
// multiply a by a link of b, return link of a
Expression ProductLink(Expression a, Expression b);
// divide a by a link of b, return link of a
Expression FractionLink(Expression a, Expression b) throw(ErrDivideByZero);
// raise a to power b, return link of a
Expression PowerLink(Expression a, Expression b);
// change sign to coeff of a, return link of a
Expression MinusLink(Expression a);
// return link to log(a)
Expression LogLink(Expression a) throw(ErrNotPermitted);
// return link to exp(a)
Expression ExpLink(Expression a);
// other unary functions (of a):
Expression SinLink(Expression a);
Expression CosLink(Expression a);
Expression TanLink(Expression a);
Expression CotLink(Expression a) throw(ErrNotPermitted);
Expression SinhLink(Expression a);
Expression CoshLink(Expression a);
Expression TanhLink(Expression a);
Expression CothLink(Expression a) throw(ErrNotPermitted);
Expression SqrtLink(Expression a) throw(ErrNotPermitted);

// these are equivalent to the above but they don't change the arguments
Expression operator + (Expression a, Expression b);
Expression operator - (Expression a, Expression b);
Expression operator * (Expression a, Expression b);
Expression operator / (Expression a, Expression b) throw(ErrDivideByZero);
Expression operator ^ (Expression a, Expression b);
Expression operator - (Expression a);
Expression Log(Expression a) throw(ErrNotPermitted);
Expression Exp(Expression a);
Expression Sin(Expression a);
Expression Cos(Expression a);
Expression Tan(Expression a);
Expression Cot(Expression a) throw(ErrNotPermitted);
Expression Sinh(Expression a);
Expression Cosh(Expression a);
Expression Tanh(Expression a);
Expression Coth(Expression a) throw(ErrNotPermitted);
Expression Sqrt(Expression a) throw(ErrNotPermitted);

// symbolic differentiation of a w.r.t. variable with index vi, 
// return the created expression (a is not changed)
Expression Diff(const Expression& a, Int vi);
Expression DiffNoSimplify(const Expression& ac, Int vi);

// SIMPLIFICATIONS - all simplifications return true if they were
// effective or false if they weren't

// sin^2+cos^2 = 1 simplification 
bool TrigSimp(Expression a);

// generic simplification with modification of the expression
bool Simplify(Expression* a);
bool SimplifyRecursive(Expression* a);
bool SimplifyConstant(Expression* a);
bool DifferenceToSum(Expression* a);
bool CompactLinearPart(Expression* a);
bool CompactLinearPartRecursive(Expression* a);
bool CompactProducts(Expression* a);
bool ReorderNodes(Expression* a);

// destroys the whole tree and all nodes, be careful
void RecursiveDestroy(Expression* a);

// generic simplification on a copy of the expression
Expression SimplifyCopy(Expression* a, bool& ischanged);

  
#endif
